File: src/main/java/me/glaremasters/guilds/arena/Arena.java
package me.glaremasters.guilds.arena;

import co.aikar.commands.ACFBukkitUtil;
import org.bukkit.Location;

import java.util.UUID;

public class Arena {

    private UUID id;
    private String name;
    private String challenger;
    private String defender;
    private transient boolean inUse;

    public Arena(UUID id, String name) {
        this.id = id;
        this.name = name;
    }

    public Arena(UUID id, String name, String challenger, String defender, boolean inUse) {
        this.id = id;
        this.name = name;
        this.challenger = challenger;
        this.defender = defender;
        this.inUse = inUse;
    }

    public static ArenaBuilder builder() {
        return new ArenaBuilder();
    }

    /**
     * Get the challenger as a location object
     * @return loc object
     */
    public Location getChallengerLoc() {
        return ACFBukkitUtil.stringToLocation(challenger);
    }

    /**
     * Get the defender as a location object
     * @return loc object
     */
    public Location getDefenderLoc() {
        return ACFBukkitUtil.stringToLocation(defender);
    }

    public UUID getId() {
        return this.id;
    }

    public String getName() {
        return this.name;
    }

    public String getChallenger() {
        return this.challenger;
    }

    public String getDefender() {
        return this.defender;
    }

    public boolean isInUse() {
        return this.inUse;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setChallenger(String challenger) {
        this.challenger = challenger;
    }

    public void setDefender(String defender) {
        this.defender = defender;
    }

    public void setInUse(boolean inUse) {
        this.inUse = inUse;
    }

    public static class ArenaBuilder {
        private UUID id;
        private String name;
        private String challenger;
        private String defender;
        private boolean inUse;

        ArenaBuilder() {
        }

        public Arena.ArenaBuilder id(UUID id) {
            this.id = id;
            return this;
        }

        public Arena.ArenaBuilder name(String name) {
            this.name = name;
            return this;
        }

        public Arena.ArenaBuilder challenger(String challenger) {
            this.challenger = challenger;
            return this;
        }

        public Arena.ArenaBuilder defender(String defender) {
            this.defender = defender;
            return this;
        }

        public Arena.ArenaBuilder inUse(boolean inUse) {
            this.inUse = inUse;
            return this;
        }

        public Arena build() {
            return new Arena(id, name, challenger, defender, inUse);
        }

        public String toString() {
            return "Arena.ArenaBuilder(id=" + this.id + ", name=" + this.name + ", challenger=" + this.challenger + ", defender=" + this.defender + ", inUse=" + this.inUse + ")";
        }
    }
}

File: src/main/kotlin/me/glaremasters/guilds/arena/Arena.kt
/*
 * MIT License
 *
 * Copyright (c) 2023 Glare
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package me.glaremasters.guilds.arena

import co.aikar.commands.ACFBukkitUtil
import java.util.UUID
import org.bukkit.Location

/**
 * Represents an Arena with unique id and name.
 *
 * @property id the unique id of the arena.
 * @property name the name of the arena.
 * @property challenger the challenger location string representation, can be null.
 * @property defender the defender location string representation, can be null.
 * @property inUse boolean flag indicating whether the arena is currently in use.
 */
data class Arena(
    val id: UUID,
    val name: String,
    var challenger: String?,
    var defender: String?,
    @Transient var inUse: Boolean
) {

    /**
     * Constructs a new Arena with id and name.
     *
     * @param id the unique id of the arena.
     * @param name the name of the arena.
     */
    constructor(id: UUID, name: String) : this(id, name, null, null, false)

    /**
     * Returns the Location object for the challenger.
     *
     * @return the Location object for the challenger, null if the string representation is null.
     */
    val challengerLoc: Location?
        get() = ACFBukkitUtil.stringToLocation(challenger)

    /**
     * Returns the Location object for the defender.
     *
     * @return the Location object for the defender, null if the string representation is null.
     */
    val defenderLoc: Location?
        get() = ACFBukkitUtil.stringToLocation(defender)
}
